core/src/main/java/org/mskcc/oncotree/crosswalk/MSKConceptCache.java

```diff
@@ -67,20 +70,38 @@ public MSKConcept get(String oncoTreeCode) {
     private void resetCache() {
         logger.info("resetCache() -- clearing Crosswalk MSKConcept cache and refilling");
         oncoTreeCodesToMSKConcepts.clear();
-        List<OncoTreeNode> oncoTreeNodes = oncoTreeRepository.getOncoTree(VersionUtil.getDefaultVersion());
-        for (OncoTreeNode node : oncoTreeNodes) {
-            getFromCrosswalkAndSave(node.getCode());
-        } 
+        // versions are ordered in ascending order by release date
+        for (Version version : VersionUtil.getVersions()) {
+            List<OncoTreeNode> oncoTreeNodes = oncoTreeRepository.getOncoTree(version);
+            for (OncoTreeNode node : oncoTreeNodes) {
+                MSKConcept mskConcept = getFromCrosswalkAndSave(node.getCode());
+                // get all codes defined so far for this topbraid uri and save in history
+                if (topBraidURIsToOncotreeCodes.containsKey(node.getURI())) {
+                    // do not add this code to the history, but add any others
+                    HashSet<String> allButThisNode = new HashSet<String>(topBraidURIsToOncotreeCodes.get(node.getURI()));
+                    allButThisNode.remove(node.getCode());
+                    mskConcept.addHistory(allButThisNode);
+                } else {
+                    topBraidURIsToOncotreeCodes.put(node.getURI(), new HashSet<String>());
+                }
+                // now save this as onoctree code history for this topbraid uri
+                topBraidURIsToOncotreeCodes.get(node.getURI()).add(node.getCode());
+            }
+        }
     }
 
     private MSKConcept getFromCrosswalkAndSave(String oncoTreeCode) {
-        MSKConcept concept = null;
+        // only save if we have not seen before (UMLS/NCI info will not be different)
+        if (oncoTreeCodesToMSKConcepts.containsKey(oncoTreeCode)) {
+            return oncoTreeCodesToMSKConcepts.get(oncoTreeCode);
```

Comment:
ah .. maybe this covers the "returned to original code" case I mentioned above. But it also may "lose" the information from the "version 2" and "version 3" history.